engine/src/retrier.rs

```diff
@@ -269,22 +270,32 @@ impl<Client: Send + Sync + Clone + 'static> ClientSelector<Client> {
 			None
 		};
 
-		Self { primary_signal, secondary_signal }
+		Self {
+			primary_signal,
+			secondary_signal,
+			prefer: PrimaryOrSecondary::Primary,
+			last_tried_primary: None,
+		}
 	}
 
 	// Returns a client, and the type of client selected.
-	pub async fn select_client(
-		&self,
-		primary_or_secondary: PrimaryOrSecondary,
-	) -> (Client, PrimaryOrSecondary) {
+	pub async fn select_client(&mut self) -> (Client, PrimaryOrSecondary) {
 		futures::future::select_all(
 			utilities::conditional::conditional(
 				&self.secondary_signal,
 				|secondary_signal| {
 					// If we have two clients, then we should bias the requested one, but if it's
 					// not ready, request from the other one.
-					match primary_or_secondary {
-						PrimaryOrSecondary::Secondary => [secondary_signal, &self.primary_signal],
+					match self.prefer {
```

Comment:
You don't truely need this `prefer` member, you could only have the option and set it to None, once the duration has past. If you did this you could us a [Timer](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) instead of a Instant.

Reply:
I see this would slightly change the behaviour actually so probably not worth it, but it does seem odd to have both members.

Reply:
Agreed, that it seems odd, thought about it for a bit. But settled for this as the intention is at least clear.